massa-versioning-worker/src/versioning_ser_der.rs

```diff
@@ -170,10 +179,463 @@ impl Deserializer<MipInfo> for MipInfoDeserializer {
     }
 }
 
+// End MipInfo
+
+// ComponentState
+
+/// Serializer for `ComponentState`
+#[derive(Clone)]
+pub struct ComponentStateSerializer {
+    u32_serializer: U32VarIntSerializer,
+    amount_serializer: AmountSerializer,
+}
+
+impl ComponentStateSerializer {
+    /// Creates a new `Serializer`
+    pub fn new() -> Self {
+        Self {
+            u32_serializer: U32VarIntSerializer::new(),
+            amount_serializer: AmountSerializer::new(),
+        }
+    }
+}
+
+impl Default for ComponentStateSerializer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Serializer<ComponentState> for ComponentStateSerializer {
+    fn serialize(
+        &self,
+        value: &ComponentState,
+        buffer: &mut Vec<u8>,
+    ) -> Result<(), SerializeError> {
+        let (state, threshold_): (u32, Option<Amount>) = match value {
+            ComponentState::Error => (u32::from(ComponentStateTypeId::Error), None),
+            ComponentState::Defined(_) => (u32::from(ComponentStateTypeId::Defined), None),
+            ComponentState::Started(Started { threshold }) => {
+                (u32::from(ComponentStateTypeId::Started), Some(*threshold))
+            }
+            ComponentState::LockedIn(_) => (u32::from(ComponentStateTypeId::LockedIn), None),
+            ComponentState::Active(_) => (u32::from(ComponentStateTypeId::Active), None),
+            ComponentState::Failed(_) => (u32::from(ComponentStateTypeId::Failed), None),
+        };
+        self.u32_serializer.serialize(&state, buffer)?;
+        if let Some(threshold) = threshold_ {
+            self.amount_serializer.serialize(&threshold, buffer)?;
+        }
+        Ok(())
+    }
+}
+
+/// A Deserializer for ComponentState`
+pub struct ComponentStateDeserializer {
+    state_deserializer: U32VarIntDeserializer,
+    amount_deserializer: AmountDeserializer,
+}
+
+impl ComponentStateDeserializer {
+    /// Creates a new ``
+    pub fn new() -> Self {
+        Self {
+            state_deserializer: U32VarIntDeserializer::new(
+                Included(0),
+                Excluded(COMPONENT_STATE_VARIANT_COUNT + 1),
```

Comment:
If the +1 is to account for the Error, why don't we use `COMPONENT_STATE_ID_VARIANT_COUNT` here?
More generally, I think having two different counts may be prone to errors (especially has they differ by just one, and the Error state is added automatically!).
At the very least, needs more documentations.

Reply:
Good catch! I've removed this +1 :)

Reply:
Ok great!